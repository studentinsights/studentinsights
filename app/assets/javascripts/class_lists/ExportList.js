import React from 'react';
import {Creatable} from 'react-select';
import 'react-select/dist/react-select.css';
import _ from 'lodash';
import DownloadCsvLink, {joinCsvRow} from '../components/DownloadCsvLink';
import SuccessLabel from '../components/SuccessLabel';
import {gradeText} from '../helpers/gradeText';
import {
  UNPLACED_ROOM_KEY,
  createRooms,
  findMovedStudentIds
} from './studentIdsByRoomFunctions';


// Allows principals to set the names of teachers for each room, and export
// to a spreadsheet.  Others can view but can't export if the principal hasn't
// added names yet.
// The CSV download is within the browser.
export default class ExportList extends React.Component {
  mapToRows(studentIdsByRoom) {
    const {gradeLevelNextYear, students, principalTeacherNamesByRoom} = this.props;

    const rows = _.flatten(_.compact(Object.keys(studentIdsByRoom).map(roomKey => {
      return studentIdsByRoom[roomKey].map(studentId => {
        const student = _.find(students, {id: studentId});
        const teacherText = principalTeacherNamesByRoom[roomKey] || 'Not placed';
        return [
          `${student.last_name}, ${student.first_name}`,
          student.local_id,
          gradeLevelNextYear,
          teacherText
        ];
      });
    })));

    // Sort by name for secretaries
    return _.sortBy(rows, row => row[0]);
  }

  isReadyToExport(studentIdsByRoom) {
    const {principalTeacherNamesByRoom} = this.props;
    const isMissingTeacherNames = (Object.keys(principalTeacherNamesByRoom).length < Object.keys(studentIdsByRoom).length - 1);
    const hasBlankTeacherName = _.any(_.values(principalTeacherNamesByRoom), _.isEmpty);

    return (!isMissingTeacherNames && !hasBlankTeacherName);
  }

  // This overrides the default, which includes the comma key as well, which
  // doesn't make sense with (lastname, firstname) here.
  shouldKeyDownEventCreateNewOption(keyCode) {
    if ((keyCode === 9) || (keyCode === 13)) return true;
    return false;
  }

  onRoomTeacherChanged(roomKey, option) {
    const {principalTeacherNamesByRoom, onPrincipalTeacherNamesByRoomChanged} = this.props;
    onPrincipalTeacherNamesByRoomChanged({
      ...principalTeacherNamesByRoom,
      [roomKey]: (option === null) ? '' : option.name
    });
  }

  render() {
    const {
      headingStyle,
      teacherStudentIdsByRoom,
      principalStudentIdsByRoom,
      gradeLevelNextYear
    } = this.props;
    const studentIdsByRoom = principalStudentIdsByRoom || teacherStudentIdsByRoom; // TODO(kr)
    const rows = this.mapToRows(studentIdsByRoom);

    return (
      <div className="ExportList">
        <div style={headingStyle}>Have all students been placed for next year's {gradeText(gradeLevelNextYear)}?</div>
        <div style={styles.spaceBelow}>{this.renderUnplaced(studentIdsByRoom)}</div>
        
        <div style={headingStyle}>Have you reviewed the team's plan and notes before making changes?</div>
        <div style={styles.spaceBelow}>{this.renderMoved(studentIdsByRoom)}</div>

        <div style={headingStyle}>Who will teach next year's {gradeText(gradeLevelNextYear)}?</div>
        <div>{this.renderTeacherAssignment(studentIdsByRoom)}</div>
        <div>{this.renderDownloadListsLink(studentIdsByRoom, rows)}</div>
      </div>
    );
  }

  renderUnplaced(studentIdsByRoom) {
    const unplacedCount = studentIdsByRoom[UNPLACED_ROOM_KEY].length;
    if (unplacedCount === 0) return <SuccessLabel style={styles.placementMessage} text="All students have been placed." />;
    if (unplacedCount === 1) return this.renderWarning(`There is one student who has not been placed in a classroom.`);
    if (unplacedCount > 1) return this.renderWarning(`There are ${unplacedCount} students who have not been placed in a classroom.`);
  }

  renderMoved(studentIdsByRoom) {
    const {teacherStudentIdsByRoom} = this.props;
    const movedStudentsCount = findMovedStudentIds(teacherStudentIdsByRoom, studentIdsByRoom).length;

    if (movedStudentsCount === 0) return <SuccessLabel style={styles.placementMessage} text="No students were moved." />;
    if (movedStudentsCount === 1) return this.renderWarning(`You made one revision to the team's original plan.`);
    if (movedStudentsCount > 1) return this.renderWarning(`You made ${movedStudentsCount} revisions to the team's original plan.`);
  }

  renderWarning(text) {
    return <SuccessLabel style={{...styles.placementMessage, ...styles.placementWarning}} text={text} />;
  }

  renderTeacherAssignment(studentIdsByRoom) {
    const {principalTeacherNamesByRoom, educators, isRevisable} = this.props;
    const rooms = createRooms(Object.keys(studentIdsByRoom).length - 1).filter(room => {
      return room.roomKey !== UNPLACED_ROOM_KEY;
    });

    // Include created names as options as well.
    const createdNames = _.values(principalTeacherNamesByRoom);
    const educatorNames = educators.map(educator => educator.full_name);
    const uniqueSortedNames = _.sortBy(_.uniq(createdNames.concat(educatorNames)), educator => educator.toLowerCase());
    const options = uniqueSortedNames.map(name => { return {name}; });

    return (
      <table style={styles.table}>
        <tbody>{rooms.map(room => {
          const teacherText = principalTeacherNamesByRoom[room.roomKey] || '';
          return (
            <tr key={room.roomKey}>
              <td style={styles.cell}>{room.roomName}</td>
              <td style={styles.cell}>
                <Creatable
                  valueKey="name"
                  labelKey="name"
                  placeholder="Select or type..."
                  style={styles.select}
                  options={options}
                  disabled={!isRevisable}
                  autosize={false} // IE11, see https://github.com/JedWatson/react-select/issues/733#issuecomment-237562382
                  multi={false}
                  clearable={false}
                  value={teacherText}
                  promptTextCreator={_.identity}
                  shouldKeyDownEventCreateNewOption={this.shouldKeyDownEventCreateNewOption}
                  onChange={this.onRoomTeacherChanged.bind(this, room.roomKey)}
                  backspaceRemoves={true} />
              </td>
            </tr>
          );
        })}</tbody>
      </table>
    );
  }

  renderDownloadListsLink(studentIdsByRoom, rows) {
    const {nowFn} = this.context;
    const now = nowFn();
    const isReadyToExport = this.isReadyToExport(studentIdsByRoom);
    const {gradeLevelNextYear, school} = this.props;
    const gradeLevelText = gradeText(gradeLevelNextYear);
    const dateText = now.format('YYYY-MM-DD');
    const filename = `Class list: ${gradeLevelText} at ${school.name} ${dateText}.csv`;
    const header = joinCsvRow([
      'Student name',
      'LASID',
      'Grade level next year',
      'Room next year'
    ]);
    const csvText = [header].concat(rows.map(joinCsvRow)).join('\n');

    return (
      <div>
        <DownloadCsvLink
          disabled={!isReadyToExport}
          style={styles.download}
          filename={filename}
          csvText={csvText}>
          Download class lists spreadsheet
        </DownloadCsvLink>
        {!isReadyToExport &&
          <div style={styles.warnExport}>{this.renderWarning('Teachers need to be assigned to each homeroom first')}</div>
        }
      </div>
    );
  }
}
ExportList.contextTypes = {
  nowFn: React.PropTypes.func.isRequired
};
ExportList.propTypes = {
  isRevisable: React.PropTypes.bool.isRequired,
  gradeLevelNextYear: React.PropTypes.string.isRequired,
  school: React.PropTypes.shape({
    name: React.PropTypes.string.isRequired
  }),
  students: React.PropTypes.arrayOf(React.PropTypes.shape({
    id: React.PropTypes.number.isRequired,
    local_id: React.PropTypes.string.isRequired,
    first_name: React.PropTypes.string.isRequired,
    last_name: React.PropTypes.string.isRequired
  })).isRequired,
  fetchProfile: React.PropTypes.func.isRequired,
  teacherStudentIdsByRoom: React.PropTypes.object.isRequired,
  principalStudentIdsByRoom: React.PropTypes.object,
  educators: React.PropTypes.arrayOf(React.PropTypes.shape({
    full_name: React.PropTypes.string.isRequired
  })).isRequired,
  principalTeacherNamesByRoom: React.PropTypes.object.isRequired,
  onPrincipalTeacherNamesByRoomChanged:  React.PropTypes.func,
  headingStyle: React.PropTypes.object
};


const styles = {
  placementMessage: {
    display: 'inline-block',
    marginTop: 10,
    marginBottom: 10
  },
  placementWarning: {
    background: 'orange',
    color: 'white',
    borderColor: 'darkorange'
  },
  select: {
    width: '20em'
  },
  table: {
    borderCollapse: 'collapse',
    marginTop: 5
  },
  cell: {
    padding: '2px 8px'
  },
  download: {
    display: 'inline-block',
    textDecoration: 'none',
    color: 'white',
    marginTop: 20
  },
  spaceBelow: {},
  warnExport: {
    display: 'inline-block',
    marginLeft: 10
  }
};
