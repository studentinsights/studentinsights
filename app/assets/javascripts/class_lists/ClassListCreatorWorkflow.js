import React from 'react';
import _ from 'lodash';
import Select from 'react-select';
import 'react-select/dist/react-select.css';
import {gradeText} from '../helpers/gradeText';
import Loading from '../components/Loading';
import {SeriousButton} from '../components/Button';
import SuccessLabel from '../components/SuccessLabel';
import IntroCopy from './IntroCopy';
import CreateYourLists, {styleStudentFn} from './CreateYourLists';
import ExportList from './ExportList';
import HorizontalStepper from './HorizontalStepper';
import {fetchProfile} from './api';
import {resolveDriftForStudents, findMovedStudentIds} from './studentIdsByRoomFunctions';



// This is the UI component for grade-level teaching teams to go through the process
// of creating classroom lists.  It tracks all changes and passes them up to callbacks,
// and hands off to other UI components that handle stepping through the process
// and screens for each step.
export default class ClassListCreatorWorkflow extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      isExpandedVertically: false
    };
    this.renderStepContents = this.renderStepContents.bind(this);
    this.onExpandVerticallyToggled = this.onExpandVerticallyToggled.bind(this);
  }

  onExpandVerticallyToggled() {
    const {isExpandedVertically} = this.state;
    this.setState({isExpandedVertically: !isExpandedVertically});
  }

  render() {
    const {
      steps,
      stepIndex,
      availableSteps,
      onStepChanged,
      isRevisable,
      isEditable,
      isDirty
    } = this.props;
    const {isExpandedVertically} = this.state;
    const expandedOrCollapsedStyles = (isExpandedVertically)
      ? styles.horizontalStepperExpanded 
      : styles.horizontalStepperCollapsed;
    return (
      <div className="ClassListCreatorView" style={styles.root}>
        <HorizontalStepper
          steps={steps}
          availableSteps={availableSteps}
          isEditable={isEditable}
          isRevisable={isRevisable}
          isDirty={isDirty}
          stepIndex={stepIndex}
          onStepChanged={onStepChanged}
          renderFn={this.renderStepContents}
          style={{...styles.horizontalStepper, ...expandedOrCollapsedStyles}}
          contentStyle={styles.horizontalStepperContent} />
      </div>
    );
  }

  renderStepContents(stepIndex, step) {
    if (stepIndex === 0) return this.renderChooseYourGrade();
    if (stepIndex === 1) return this.renderMakeAPlan();
    if (stepIndex === 2) return this.renderCreateYourClassrooms();
    if (stepIndex === 3) return this.renderSubmit();
    if (stepIndex === 4) return this.renderPrincipalFinalizes();
    if (stepIndex === 5) return this.renderExportList();
  }

  renderChooseYourGrade() {
    const {
      isEditable,
      schools,
      gradeLevelsNextYear,
      schoolId,
      gradeLevelNextYear,
      onSchoolIdChanged,
      canChangeSchoolOrGrade,
      onGradeLevelNextYearChanged
    } = this.props;

    if (schools === null || gradeLevelsNextYear === null) return <Loading />;

    return (
      <div key="choose-your-grade" style={styles.stepContent}>
        <div>
          <div style={styles.titleHeading}>Class List Creator</div>
          <IntroCopy />
        </div>
        <div>
          <div>
            <div style={styles.heading}>What school?</div>
              <Select
                name="select-school-name"
                value={schoolId}
                onChange={item => onSchoolIdChanged(item.value)}
                disabled={!isEditable || !canChangeSchoolOrGrade}
                options={_.sortBy(schools, s => s.name).map(school => {
                  return {
                    value: school.id,
                    label: school.name
                  };
                })}
              />
          </div>
          <div>
            <div style={styles.heading}>What grade level are you creating?</div>
              <Select
                name="select-grade-level"
                value={gradeLevelNextYear}
                onChange={item => onGradeLevelNextYearChanged(item.value)}
                disabled={!isEditable || !canChangeSchoolOrGrade}
                options={gradeLevelsNextYear.map(gradeLevelNextYear => {
                  return {
                    value: gradeLevelNextYear,
                    label: `Next year's ${gradeText(gradeLevelNextYear)} ` 
                  };
                })}
              />
          </div>
          {!canChangeSchoolOrGrade &&
            <div style={{marginTop: 20}}>You can't change the school or grade level once you've moved forward.  If you need to change this, <a href="/classlists">create a new class list</a> instead.</div>
          }
        </div>
      </div>
    );
  }

  renderMakeAPlan() {
    const {
      isEditable,
      educators,
      authors,
      students,
      gradeLevelNextYear,
      classroomsCount,
      planText,
      onEducatorsChanged,
      onClassroomsCountIncremented,
      onPlanTextChanged
    } = this.props;

    if (educators === null || students === null) return <Loading />;
    return (
      <div key="make-a-plan" style={styles.stepContent}>
        <div>
          <div style={styles.heading}>Who's the team creating these class lists?</div>
          <Select
            name="select-educators"
            multi
            removeSelected
            value={authors}
            valueKey="id"
            labelKey="full_name"
            options={educators}
            onChange={onEducatorsChanged}
            disabled={!isEditable}
          />
        </div>
        <div>
          <div style={styles.heading}>How many {gradeText(gradeLevelNextYear)} classrooms will you create?</div>
          <div style={{marginLeft: 5, display: 'inline-block'}}>
            <button
              style={styles.incrementButton}
              disabled={classroomsCount <= 2 || !isEditable}
              onClick={() => onClassroomsCountIncremented(-1)}>
              -
            </button>
            <div style={{display: 'inline-block', padding: 10}}>{classroomsCount} classrooms</div>
            <button
              style={styles.incrementButton}
              disabled={classroomsCount >= 5 || !isEditable}
              onClick={() => onClassroomsCountIncremented(1)}>
              +
            </button>
          </div>
          <div style={{display: 'inline-block', fontSize: 12, marginLeft: 20}}>With {students.length} students total, this makes the <b>average class size {Math.ceil(students.length / classroomsCount)} students</b>.</div>
        </div>
        <div>
          <div style={styles.heading}>What's your plan for creating classroom communitites?</div>
          <div style={{fontSize: 12, padding: 10, paddingLeft: 0, paddingTop: 3}}>
            Some teams start with considering social dynamics, splitting up students who are leaders or who don't work well together.  Others start creating groups with diverse academic strengths.
          </div>
          <div>
            <textarea
              style={styles.textarea}
              disabled={!isEditable}
              rows={8}
              value={planText}
              onChange={event => onPlanTextChanged(event.target.value)} />
          </div>
        </div>
      </div>
    );
  }

  renderCreateYourClassrooms() {    
    const {
      workspaceId,
      isEditable,
      students,
      classroomsCount,
      onClassListsChanged,
      studentIdsByRoom,
      gradeLevelNextYear
    } = this.props;
    const {isExpandedVertically} = this.state;

    if (students === null || studentIdsByRoom === null) return <Loading />;
    return (
      <CreateYourLists
        key="create-your-classrooms"
        students={students}
        classroomsCount={classroomsCount}
        gradeLevelNextYear={gradeLevelNextYear}
        studentIdsByRoom={resolveDriftForStudents(studentIdsByRoom, _.map(students, 'id'))}
        fetchProfile={studentId => fetchProfile(workspaceId, studentId)}
        isEditable={isEditable}
        isExpandedVertically={isExpandedVertically}
        onExpandVerticallyToggled={this.onExpandVerticallyToggled}
        onClassListsChanged={onClassListsChanged}/>
    );
  }

  renderSubmit() {
    const {
      isEditable,
      isSubmitted,
      isDirty,
      onPrincipalNoteChanged,
      principalNoteText,
      onFeedbackTextChanged,
      feedbackText,
      onSubmitClicked
    } = this.props;

    return (
      <div key="submit" style={styles.stepContent}>
        <div>
          <div>What else should your principal know?</div>
          <div style={styles.descriptionText}>
            Putting in these notes will help your principal and other team members understand all the different factors that you considered besides what shows up in the graphs.  This is also crucial information for a principal to know in case they need to move any students around over the summer.
          </div>
          <textarea
            value={principalNoteText}
            disabled={!isEditable}
            onChange={event => onPrincipalNoteChanged(event.target.value)}
            rows={5} 
            style={styles.textarea} />
        </div>
        <div style={styles.marginBetweenSections}>
          <div>Any feedback?</div>
          <div style={styles.descriptionText}>
            Let us know any feedback you have so we can improve this for next year.
          </div>
          <textarea
            value={feedbackText}
            disabled={!isEditable}
            onChange={event => onFeedbackTextChanged(event.target.value)}
            rows={2} 
            style={styles.textarea} />
        </div>
        <div style={styles.marginBetweenSections}>
          <div>Submit</div>
          <div style={{display: 'flex', justifyContent: 'space-between'}}>
            <div style={styles.descriptionText}>After you submit your class list, the principal will be the only one who can make changes.</div>
            {isSubmitted
              ? (isDirty) ? <span>Saving...</span> : <SuccessLabel text="Your class list is submitted" />
              : <SeriousButton isDisabled={!isEditable} onClick={onSubmitClicked}>Submit to principal</SeriousButton>
            }
          </div>
        </div>
      </div>
    );
  }

  renderPrincipalFinalizes() {
    const {
      workspaceId,
      students,
      isRevisable,
      classroomsCount,
      gradeLevelNextYear,
      onClassListsChangedByPrincipal
    } = this.props;
    const {isExpandedVertically} = this.state;

    // Default to teacher lists if no moves
    const teacherStudentIdsByRoom = this.props.studentIdsByRoom;
    const principalStudentIdsByRoom = this.props.principalStudentIdsByRoom || teacherStudentIdsByRoom;
    if (students === null || principalStudentIdsByRoom === null) return <Loading />;

    // See which students have moved so we can style them differently.
    const movedStudentIds = findMovedStudentIds(teacherStudentIdsByRoom, principalStudentIdsByRoom);
    return (
      <CreateYourLists
        key="principal-finalizes"
        isEditable={isRevisable}
        students={students}
        classroomsCount={classroomsCount}
        gradeLevelNextYear={gradeLevelNextYear}
        studentIdsByRoom={resolveDriftForStudents(principalStudentIdsByRoom, _.map(students, 'id'))}
        fetchProfile={studentId => fetchProfile(workspaceId, studentId)}
        styleStudentFn={student => styleStudentFn(movedStudentIds, student)}
        isExpandedVertically={isExpandedVertically}
        onExpandVerticallyToggled={this.onExpandVerticallyToggled}
        onClassListsChanged={onClassListsChangedByPrincipal} />
    );
  }

  renderExportList() {
    const {
      workspaceId,
      isRevisable,
      gradeLevelNextYear,
      schoolId,
      schools,
      students,
      studentIdsByRoom,
      principalTeacherNamesByRoom,
      educators,
      principalStudentIdsByRoom,
      onPrincipalTeacherNamesByRoomChanged
    } = this.props;
    const school = _.find(schools, {id: schoolId});

    if (students === null || educators === null || studentIdsByRoom === null) return <Loading />;

    return (
      <div key="export" style={styles.stepContent}>
        <ExportList
          isRevisable={isRevisable}
          headingStyle={styles.heading}
          school={school}
          gradeLevelNextYear={gradeLevelNextYear}
          students={students} 
          fetchProfile={studentId => fetchProfile(workspaceId, studentId)}
          teacherStudentIdsByRoom={studentIdsByRoom}
          principalStudentIdsByRoom={principalStudentIdsByRoom}
          educators={educators}
          principalTeacherNamesByRoom={principalTeacherNamesByRoom}
          onPrincipalTeacherNamesByRoomChanged={isRevisable ? onPrincipalTeacherNamesByRoomChanged : null}
        />
      </div>
    );
  }
}
ClassListCreatorWorkflow.propTypes = {
  // server data
  schools: React.PropTypes.array,
  gradeLevelsNextYear: React.PropTypes.array,
  students: React.PropTypes.array,
  educators: React.PropTypes.array,

  // config
  steps: React.PropTypes.arrayOf(React.PropTypes.string).isRequired,
  availableSteps: React.PropTypes.arrayOf(React.PropTypes.number).isRequired,
  isEditable: React.PropTypes.bool.isRequired,
  isSubmitted: React.PropTypes.bool.isRequired,
  isRevisable: React.PropTypes.bool.isRequired,

  // workspace
  isDirty: React.PropTypes.bool.isRequired,
  canChangeSchoolOrGrade: React.PropTypes.bool.isRequired,
  stepIndex: React.PropTypes.number.isRequired,
  workspaceId: React.PropTypes.string.isRequired,
  schoolId: React.PropTypes.number,
  gradeLevelNextYear: React.PropTypes.string,
  authors: React.PropTypes.array.isRequired,
  classroomsCount: React.PropTypes.number.isRequired,
  planText: React.PropTypes.string.isRequired,
  studentIdsByRoom: React.PropTypes.object,
  principalNoteText: React.PropTypes.string.isRequired,
  feedbackText: React.PropTypes.string.isRequired,

  // principal
  principalTeacherNamesByRoom: React.PropTypes.object,
  principalStudentIdsByRoom: React.PropTypes.object,
  onClassListsChangedByPrincipal: React.PropTypes.func.isRequired,
  onPrincipalTeacherNamesByRoomChanged: React.PropTypes.func.isRequired,

  // callbacks
  onStepChanged: React.PropTypes.func.isRequired,
  onSchoolIdChanged: React.PropTypes.func.isRequired,
  onGradeLevelNextYearChanged: React.PropTypes.func.isRequired,
  onEducatorsChanged: React.PropTypes.func.isRequired,
  onClassroomsCountIncremented: React.PropTypes.func.isRequired,
  onPlanTextChanged: React.PropTypes.func.isRequired,
  onClassListsChanged: React.PropTypes.func.isRequired,
  onPrincipalNoteChanged: React.PropTypes.func.isRequired,
  onFeedbackTextChanged: React.PropTypes.func.isRequired,
  onSubmitClicked: React.PropTypes.func.isRequired
};

const styles = {
  root: {
    fontSize: 14,
    width: '100%'
  },
  heading: {
    marginTop: 20
  },
  titleHeading: {
    fontSize: 20,
    fontWeight: 300,
    marginTop: 20,
    marginBottom: 10
  },
  button: {
    display: 'inline-block',
    margin: 5,
    cursor: 'pointer'
  },
  incrementButton: {
    display: 'inline-block',
    padding: 1,
    paddingLeft: 10,
    paddingRight: 10,
    fontSize: 16
  },
  stepContent: {
    margin: 20
  },
  videoLink: {
    display: 'inline-block',
    marginLeft: 5,
  },
  horizontalStepper: {
    paddingTop: 15
  },
  horizontalStepperCollapsed: {
    height: '100%',
    display: 'flex',
    flexDirection: 'column'
  },
  horizontalStepperExpanded: {
    
  },
  horizontalStepperContent: {
    borderTop: '1px solid #ccc',
    marginTop: 10
  },
  textarea: {
    border: '1px solid #ccc',
    width: '100%'
  },
  descriptionText: {
    paddingTop: 5,
    padding: 10,
    paddingLeft: 0,
    fontSize: 12
  },
  marginBetweenSections: {
    marginTop: 20
  }
};
