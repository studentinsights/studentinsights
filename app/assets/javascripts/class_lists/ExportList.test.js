import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import {withDefaultNowContext} from '../../../../spec/javascripts/support/NowContainer';
import ExportList from './ExportList';
import students_for_grade_level_next_year_json from './fixtures/students_for_grade_level_next_year_json';
import {exportPropsWithTeacherNames} from './ClassListCreatorWorkflow.test';


export function testProps(props = {}) {
  const workflowProps = exportPropsWithTeacherNames();
  return {
    ...workflowProps,
    school: {
      name: 'Awesome school'
    },
    teacherStudentIdsByRoom: workflowProps.studentIdsByRoom,
    educators: students_for_grade_level_next_year_json.educators,
    fetchProfile: function() {},
    ...props
  };
}

export function testEl(props = {}) {
  return withDefaultNowContext(<ExportList {...props} />);
}

it('renders without crashing', () => {
  const el = document.createElement('div');
  const props = testProps();
  ReactDOM.render(testEl(props), el);
});

it('snapshots', () => {
  const props = testProps();
  const tree = renderer
    .create(testEl(props))
    .toJSON();
  expect(tree).toMatchSnapshot();
});
