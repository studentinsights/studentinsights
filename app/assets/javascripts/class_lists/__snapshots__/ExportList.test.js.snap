// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshots 1`] = `
<div
  className="ExportList"
>
  <div>
    Have all students been placed for next year’s 
    1st grade
    ?
  </div>
  <div
    style={{}}
  >
    <span
      className="SuccessLabel"
      style={
        {
          "backgroundColor": "rgb(209, 231, 210)",
          "border": "1px solid rgb(149, 198, 150)",
          "borderRadius": 3,
          "color": "green",
          "display": "inline-block",
          "fontWeight": "bold",
          "marginBottom": 10,
          "marginTop": 10,
          "padding": 8,
        }
      }
    >
      All students have been placed.
    </span>
  </div>
  <div>
    Have you reviewed the team’s plan and notes before making changes?
  </div>
  <div
    style={{}}
  >
    <span
      className="SuccessLabel"
      style={
        {
          "backgroundColor": "rgb(209, 231, 210)",
          "border": "1px solid rgb(149, 198, 150)",
          "borderRadius": 3,
          "color": "green",
          "display": "inline-block",
          "fontWeight": "bold",
          "marginBottom": 10,
          "marginTop": 10,
          "padding": 8,
        }
      }
    >
      No students were moved.
    </span>
  </div>
  <div>
    Who will teach next year’s $
    1st grade
    ?
  </div>
  <div>
    <table
      style={
        {
          "borderCollapse": "collapse",
          "marginTop": 5,
        }
      }
    >
      <tbody>
        <tr>
          <td
            style={
              {
                "padding": "2px 8px",
              }
            }
          >
            Room A
          </td>
          <td
            style={
              {
                "padding": "2px 8px",
              }
            }
          >
            <div
              className="Select has-value is-disabled is-searchable Select--single"
            >
              <div
                className="Select-control"
                onKeyDown={[Function]}
                onMouseDown={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                style={
                  {
                    "width": "20em",
                  }
                }
              >
                <div
                  className="Select-multi-value-wrapper"
                  id="react-select-5--value"
                >
                  <div
                    className="Select-value"
                  >
                    <span
                      aria-selected="true"
                      className="Select-value-label"
                      id="react-select-5--value-item"
                      role="option"
                    >
                      Kevin
                    </span>
                  </div>
                  <div
                    aria-activedescendant="react-select-5--value"
                    aria-disabled="true"
                    aria-expanded={false}
                    aria-owns=""
                    className="Select-input"
                    onBlur={[Function]}
                    onFocus={[Function]}
                    role="combobox"
                    style={
                      {
                        "border": 0,
                        "display": "inline-block",
                        "width": 1,
                      }
                    }
                    tabIndex={0}
                  />
                </div>
                <span
                  className="Select-arrow-zone"
                  onMouseDown={[Function]}
                >
                  <span
                    className="Select-arrow"
                    onMouseDown={[Function]}
                  />
                </span>
              </div>
            </div>
          </td>
        </tr>
        <tr>
          <td
            style={
              {
                "padding": "2px 8px",
              }
            }
          >
            Room B
          </td>
          <td
            style={
              {
                "padding": "2px 8px",
              }
            }
          >
            <div
              className="Select has-value is-disabled is-searchable Select--single"
            >
              <div
                className="Select-control"
                onKeyDown={[Function]}
                onMouseDown={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                style={
                  {
                    "width": "20em",
                  }
                }
              >
                <div
                  className="Select-multi-value-wrapper"
                  id="react-select-6--value"
                >
                  <div
                    className="Select-value"
                  >
                    <span
                      aria-selected="true"
                      className="Select-value-label"
                      id="react-select-6--value-item"
                      role="option"
                    >
                      Alex
                    </span>
                  </div>
                  <div
                    aria-activedescendant="react-select-6--value"
                    aria-disabled="true"
                    aria-expanded={false}
                    aria-owns=""
                    className="Select-input"
                    onBlur={[Function]}
                    onFocus={[Function]}
                    role="combobox"
                    style={
                      {
                        "border": 0,
                        "display": "inline-block",
                        "width": 1,
                      }
                    }
                    tabIndex={0}
                  />
                </div>
                <span
                  className="Select-arrow-zone"
                  onMouseDown={[Function]}
                >
                  <span
                    className="Select-arrow"
                    onMouseDown={[Function]}
                  />
                </span>
              </div>
            </div>
          </td>
        </tr>
        <tr>
          <td
            style={
              {
                "padding": "2px 8px",
              }
            }
          >
            Room C
          </td>
          <td
            style={
              {
                "padding": "2px 8px",
              }
            }
          >
            <div
              className="Select has-value is-disabled is-searchable Select--single"
            >
              <div
                className="Select-control"
                onKeyDown={[Function]}
                onMouseDown={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                style={
                  {
                    "width": "20em",
                  }
                }
              >
                <div
                  className="Select-multi-value-wrapper"
                  id="react-select-7--value"
                >
                  <div
                    className="Select-value"
                  >
                    <span
                      aria-selected="true"
                      className="Select-value-label"
                      id="react-select-7--value-item"
                      role="option"
                    >
                      Uri
                    </span>
                  </div>
                  <div
                    aria-activedescendant="react-select-7--value"
                    aria-disabled="true"
                    aria-expanded={false}
                    aria-owns=""
                    className="Select-input"
                    onBlur={[Function]}
                    onFocus={[Function]}
                    role="combobox"
                    style={
                      {
                        "border": 0,
                        "display": "inline-block",
                        "width": 1,
                      }
                    }
                    tabIndex={0}
                  />
                </div>
                <span
                  className="Select-arrow-zone"
                  onMouseDown={[Function]}
                >
                  <span
                    className="Select-arrow"
                    onMouseDown={[Function]}
                  />
                </span>
              </div>
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <div>
    <div>
      <a
        className="DownloadCsvLink btn"
        download="Class list: 1st grade at Awesome school 2018-03-13.csv"
        href="data:attachment/csv,%22Student%20name%22%2C%22LASID%22%2C%22Grade%20level%20next%20year%22%2C%22Room%20next%20year%22%0A%22Disney%2C%20Donald%22%2C%22000752%22%2C%221%22%2C%22Alex%22%0A%22Kenobi%2C%20Minnie%22%2C%22000276%22%2C%221%22%2C%22Uri%22%0A%22Pan%2C%20Chip%22%2C%22000134%22%2C%221%22%2C%22Uri%22%0A%22Skywalker%2C%20Garfield%22%2C%22111111111%22%2C%221%22%2C%22Kevin%22%0A%22Skywalker%2C%20Mowgli%22%2C%22000234%22%2C%221%22%2C%22Alex%22%0A%22Skywalker%2C%20Snow%22%2C%22000525%22%2C%221%22%2C%22Kevin%22%0A%22White%2C%20Mowgli%22%2C%22000472%22%2C%221%22%2C%22Kevin%22"
        onClick={[Function]}
        style={
          {
            "color": "white",
            "display": "inline-block",
            "marginTop": 20,
            "textDecoration": "none",
          }
        }
        target="_self"
      >
        Download class lists spreadsheet
      </a>
    </div>
  </div>
</div>
`;
